### PR TITLE
feat: updated ci that is faster and better

### PR DESCRIPTION
**Pull request by jakeyheath**
### Summary
This PR updates the code that is generated to `fogg_ci.yml` for github actions. It has been tested in a few iterations in [shared-infra](https://github.com/chanzuckerberg/shared-infra/blob/main/.github/workflows/new_ci.yml). The following new features include:

* auto fix terraform-docs and commit back to branch
* auto fix fogg apply and commit back to branch
* auto fix terraform fmt and commit back to branch
* caching fogg apply using action/cache
* only doing the above on envs/accounts/modules that have changed files
* no more code generation; changed directories are collect dynamically and run in parallel
* fewer bash scripting and more github actions (such as terraform-docs)
* no more SSH keys in secrets; all code is downloaded using a github app credentials (github app's have finer grain access controls and are easier to manage at an org-wide level)
* no more random ordering of steps based on `fogg apply`. should be the same `fogg_ci.yml` every time

### Test Plan
* Testing on shared-infra

### References

In testing on shared-infra, this has been [about 4 minutes faster](https://github.com/chanzuckerberg/shared-infra/pull/5675) than the previous iteration, per PR commit.

### CODE REVIEW

File: templates/templates/.github/workflows/fogg_ci.yml.tmpl [hunk 1]
```diff
@@ -1,86 +1,102 @@
...
 jobs:
-{{- range $idx, $testBucket := $githubActionsCI.TestBuckets }}
-  terraform_{{$idx}}:
+  fogg-apply:
     runs-on: ubuntu-latest
     steps:
-      - uses: actions/checkout@v2
-      - name: Set up Python 3.x
-        uses: actions/setup-python@v1
+      - name: Generate token
```
Comment (jakeyheath) [inline templates/templates/.github/workflows/fogg_ci.yml.tmpl:8]: generate a token using our github app to do sensitive things like clone repos

File: templates/templates/.github/workflows/fogg_ci.yml.tmpl [hunk 2]
```diff
@@ -1,86 +1,102 @@
...
-      - run: aws configure set profile.{{ $profileName }}.role_arn arn:aws:iam::{{ $profile.AccountID }}:role/{{ $profile.RoleName}}
-      - run: aws configure set profile.{{ $profileName }}.source_profile _idacct
-      - run: aws --profile {{ $profileName }} sts get-caller-identity
-{{- end }}
-      # we only run the following if there are changes in the terraform/* directory
-      # to optimize CI time
-      # can use this filter to decide whether or not to run linters or tests.
-      - uses: dorny/paths-filter@v2.2.1
+          token: {{`${{ steps.generate_token.outputs.token }}`}}
+      - name: Cache Fogg
```
Comment (jakeyheath) [inline templates/templates/.github/workflows/fogg_ci.yml.tmpl:17]: utilize ~/.fogg/cache

File: templates/templates/.github/workflows/fogg_ci.yml.tmpl [hunk 3]
```diff
@@ -1,86 +1,102 @@
...
+          key: fogg-cache-{{`${{ hashFiles('**/.fogg-version') }}`}}
+      - run: make setup
+      - run: .fogg/bin/fogg apply
+        env:
+          FOGG_GITHUBAPPTOKEN: {{`${{ steps.generate_token.outputs.token }}`}}
+      - uses: EndBug/add-and-commit@v9
+        with:
+          add: -A
+          message: |
+            commit from fogg_ci -- ran fogg apply and pushed
```
Comment (jakeyheath) [inline templates/templates/.github/workflows/fogg_ci.yml.tmpl:31]: if `fogg apply` has any changes, commit them back to the branch

File: templates/templates/.github/workflows/fogg_ci.yml.tmpl [hunk 4]
```diff
@@ -1,86 +1,102 @@
...
+          initial-fetch-depth: '1'
+          list-files: json
           filters: |
-            terraform:
-              - 'terraform/**'
-  {{- range $component := $testBucket }}
-      - name: {{ $component.Dir }}
-        if: {{ "${{ steps.filter.outputs.terraform == 'true' }}" }}
+            changed:
+              - added|modified: 'terraform/**'
```
Comment (jakeyheath) [inline templates/templates/.github/workflows/fogg_ci.yml.tmpl:45]: find all changed files in terraform/* in the last commit

File: templates/templates/.github/workflows/fogg_ci.yml.tmpl [hunk 5]
```diff
@@ -1,86 +1,102 @@
...
+      - uses: actions/github-script@v6
+        id: changedDirs
```
Comment (jakeyheath) [inline templates/templates/.github/workflows/fogg_ci.yml.tmpl:47]: take the changed files and find the directories associated with their env/account/module folder

File: templates/templates/.github/workflows/fogg_ci.yml.tmpl [hunk 6]
```diff
@@ -1,86 +1,102 @@
...
+            console.log(`changedModules: ${JSON.stringify(changedModules)}`)
+            console.log(`changedAccounts: ${JSON.stringify(changedAccounts)}`)
+            console.log(`changedEnvs: ${JSON.stringify(changedEnvs)}`)
+            core.setOutput("allChanges", allChanges)
+  lint-changed-dirs:
+    runs-on: ubuntu-latest
+    needs: find-changed-dirs
+    strategy:
+      matrix:
+        tfmodule: {{`${{ fromJson(needs.find-changed-dirs.outputs.allChanges) }}`}}
```
Comment (jakeyheath) [inline templates/templates/.github/workflows/fogg_ci.yml.tmpl:67]: for each changed dir...

File: templates/templates/.github/workflows/fogg_ci.yml.tmpl [hunk 7]
```diff
@@ -1,86 +1,102 @@
...
+      - name: Generate token
+        id: generate_token
+        uses: chanzuckerberg/github-app-token@v1.1.4
+        with:
+          app_id: {{`${{ secrets.CZI_GITHUB_HELPER_APP_ID }}`}}
+          private_key: {{`${{ secrets.CZI_GITHUB_HELPER_PK }}`}}
+      - uses: actions/checkout@v3
+        with:
+          token: {{`${{ steps.generate_token.outputs.token }}`}}
+      - name: fix terraform docs
```
Comment (jakeyheath) [inline templates/templates/.github/workflows/fogg_ci.yml.tmpl:79]: fix the terraform docs and commit it back to branch

File: templates/templates/.github/workflows/fogg_ci.yml.tmpl [hunk 8]
```diff
@@ -1,86 +1,102 @@
...
+        uses: terraform-docs/gh-actions@v1.0.0
+        if: contains(matrix.tfmodule, 'modules') # only need to make docs on the modules
+        with:
+          working-dir: {{`${{matrix.tfmodule}}`}}
+          git-push: "true"
+          template: <!-- START -->\n{{`{{ .Content }}`}}\n<!-- END -->
+          ref: {{`${{ github.event.pull_request.head.ref }}`}}
+          git-commit-message: |
+            commit from fogg_ci -- ran terraform-docs and pushed
+      - name: fix terraform fmt
```
Comment (jakeyheath) [inline templates/templates/.github/workflows/fogg_ci.yml.tmpl:89]: fix terraform fmt and commit it back to the branch

File: templates/templates/.github/workflows/fogg_ci.yml.tmpl [hunk 9]
```diff
@@ -1,86 +1,102 @@
...
+        run: |
+          cd {{`${{matrix.tfmodule}}`}}
+          make fmt
+      - uses: EndBug/add-and-commit@v9
+        with:
+          add: -A
+          message: |
+            commit from fogg_ci -- ran terraform fmt and pushed
+      - name: tflint
```
Comment (jakeyheath) [inline templates/templates/.github/workflows/fogg_ci.yml.tmpl:98]: also run tflint

**Review by fctczi — APPROVED** [review]

**Pull request merged**